shell/platform/embedder/embedder.cc

```diff
@@ -1691,6 +1691,28 @@ void FlutterEngineTraceEventInstant(const char* name) {
   fml::tracing::TraceEventInstant0("flutter", name);
 }
 
+FlutterEngineResult FlutterEnginePostPlatformThreadTask(
```

Comment:
Why would the embedder need to ask the engine to post a task to the platform thread rather than doing so directly?

Reply:
So I guess it’s safe to directly post a task onto the Windows platform task runner (with target time set to `fml::TimePoint::Now()`)?

Reply:
I'm stuck here. Can you give me a hint how to achieve that?
For the time being, I've fixed `shell/platform/embedder/embedder.h` according to your comments.

Reply:
> Can you give me a hint how to achieve that?

Achieve what? The issue is that I'm not clear on what the purpose of this API is. I.e. I don't understand the use case for doing this:

> I guess it’s safe to directly post a task onto the Windows platform task runner [...]?

Why would an embedder want to use the engine's task runner for platform thread work?

Reply:
Thanks for your quick reply. This is basically needed as the `MarkTextureFrameAvailable` is expected to be called from the platform thread (see check at
https://github.com/flutter/engine/blob/b304148d465e9d30e0543128248f0f033ada4fec/shell/common/shell.cc#L953). So I was looking for a solution to post a task onto the platform task runner.
The most obvious way was to add the `PostPlatformThreadTask` method.

Reply:
If the only call site for this method is https://github.com/flutter/engine/pull/19405/files#diff-e0363cfd79295a13710333f2e74c37d754b2477bf8312306bcc53cffdbdd0ab8R56, couldn't you do `engine_->task_runner()` and post this task there? Given that this is the same platform task runner (and you could avoid adding an extra api).

Reply:
But the task runner https://github.com/flutter/engine/blob/52901f23b4b40367efb36768e4a39a624b97f955/shell/platform/windows/task_runner_win32.h#L39 expects a `FlutterTask` (declared here: https://github.com/flutter/engine/blob/52901f23b4b40367efb36768e4a39a624b97f955/shell/platform/embedder/embedder.h#L763) as opposed to a closure?

Reply:
> This is basically needed as the `MarkTextureFrameAvailable` is expected to be called from the platform thread [...]. So I was looking for a solution to post a task onto the platform task runner.

This is what I was trying to ask in the first place though: why do you want to use the platform task runner—a Flutter engine concept—to do things on the platform thread? Why not just use platform APIs directly?

I'm assuming the use case here is a plugin; plugins know the platform thread (it's the thread they were created on, and from which they must call ~all Flutter APIs). If they are running code on other threads, then they are already using platform-specific threading APIs, so why not use them to handle the cross-thread posting?

(Contrast with the method for posting to the renderer thread, which exists because the embedder may not create that thread, and thus not have any way to access it.)

Reply:
Oh, I missed that this was being used internally. I'll apply my question above to that code then, and ask: why is the embedding handling the threading, instead of the caller?

Reply:
Because as opposed to iOS/macOS platforms, we don't have someting like GCD's `dispatch_async(dispatch_get_main_queue(), ...)` on Windows, so plugin implementers have no API for "getting back" to the main thread.

So the Client Wrapper should either expose methods for explicitly posting tasks to the platform thread or methods like `MakeTextureFrameAvailable` need to be made thread agnostic as a short-term solution.
I can't think of any scenario where the exposed Texture API would give us any benefit over no API unless the rendering-related functions can be called from "anywhere".

Reply:
Look at the [flutter-webrtc](https://github.com/flutter-webrtc/flutter-webrtc) project for example.
Their macOS video renderer implemention at https://github.com/flutter-webrtc/flutter-webrtc/blob/33277024c46327bbb04fe43bec5d2866b99097f4/common/darwin/Classes/FlutterRTCVideoRenderer.m#L210 indeed uses dispatch_async() for invoking `[registry textureFrameAvailable]` on the main thread.

Reply:
> so plugin implementers have no API for "getting back" to the main thread.

Sorry, my macOS background is ingrained enough that it's hard for me to remember that there's no concept of a generally available runloop at all. I was thinking the issue was finding the thread, and forgetting that there's no way to actually run code on it without the runloop that's owned by the runner driving that.

I think the main question then is whether to plumb this through embedder.h, or pipe it in to our task runner as @iskakaushik suggested above (adding new API there as necessary). @chinmaygarde Thoughts on the embedder.h version?

Reply:
Per chat, Chinmay prefers not adding the embedder.h API (since the render thread version is actually legacy), so we should plumb it into the Windows embedding's task runner instead.

Also, you should clearly document at both the C and the C++ wrapper levels that those specific methods can be called from any thread, but also that it's the caller's responsibility to avoid races with other Flutter API calls they make. Notable ones to call out would be:
- Unregistering a texture while it's possible that a background thread is still using it.
- Shutting down the engine while it's possible that a background thread is producing texture frames.

Reply:
I've just extended the Windows TaskRunner to allow posting callbacks directly (see https://github.com/flutter/engine/pull/19405/commits/6a8a20e6e5bf139f004dbcdc7d59a0f3dbdcae87) and refactored the texture registrar accordingly besides reverting my changes to `embedder.{h,cc}`.